ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Background.hs

```diff
@@ -0,0 +1,344 @@
+import           Ouroboros.Consensus.Protocol.Abstract
+import           Ouroboros.Consensus.Util.ThreadRegistry (forkLinked)
+
+import qualified Ouroboros.Storage.ChainDB.Impl.ImmDB as ImmDB
+import qualified Ouroboros.Storage.ChainDB.Impl.LgrDB as LgrDB
+import           Ouroboros.Storage.ChainDB.Impl.Types
+import qualified Ouroboros.Storage.ChainDB.Impl.VolDB as VolDB
+
+{-------------------------------------------------------------------------------
+  Launch background tasks
+-------------------------------------------------------------------------------}
+
+launchBgTasks
+  :: forall m blk.
+     ( MonadAsync m
+     , MonadFork  m
+     , MonadMask  m
+     , MonadTime  m
+     , MonadTimer m
+     , ProtocolLedgerView blk
+     )
+  => ChainDbEnv m blk
+  -> m ()
+launchBgTasks cdb@CDB{..} = do
+    gcSchedule <- newGcSchedule
+    gcThread   <- launch $
+      gcScheduleRunner gcSchedule $ garbageCollect cdb
+    copyThread <- launch $
+      copyToImmDBRunner cdb gcSchedule
+    lgrSnapshotThread <- launch $
+      updateLedgerSnapshotsRunner cdb
+    atomically $ writeTVar cdbBgThreads
+      [gcThread, copyThread, lgrSnapshotThread]
+  where
+    launch = forkLinked cdbThreadRegistry
+
+{-------------------------------------------------------------------------------
+  Copying blocks from the VolatileDB to the ImmutableDB
+-------------------------------------------------------------------------------}
+
+-- | Copy the blocks older than @k@ from the VolatileDB to the ImmutableDB.
+--
+-- These headers of these blocks can be retrieved by dropping the @k@ most
+-- recent blocks from the fragment stored in 'cdbChain'.
+--
+-- The copied blocks are removed from the fragment stored in 'cdbChain'.
+--
+-- This function does not remove blocks from the VolatileDB.
+--
+-- The 'SlotNo' of the tip of the ImmutableDB after copying the blocks is
+-- returned. This can be used for a garbage collection on the VolatileDB.
+--
+-- NOTE: this function would not be safe when called multiple times
+-- concurrently. To enforce thread-safety, a lock is obtained at the start of
+-- this function and released at the end. So in practice, this function can be
+-- called multiple times concurrently, but the calls will be serialised.
+--
+-- NOTE: this function /can/ run concurrently with all other functions, just
+-- not with itself.
+copyToImmDB
+  :: forall m blk.
+     ( MonadSTM   m
+     , MonadCatch m
+     , OuroborosTag (BlockProtocol blk)
+     , HasHeader blk
+     , HasHeader (Header blk)
+     )
+  => ChainDbEnv m blk
+  -> m SlotNo
+copyToImmDB CDB{..} = withCopyLock $ do
+    toCopy <- atomically $ do
+      curChain <- readTVar cdbChain
+      let nbToCopy = max 0 (AF.length curChain - fromIntegral k)
+          toCopy :: [Point blk]
+          toCopy = map headerPoint
+                 $ take nbToCopy
+                 $ AF.toOldestFirst curChain
+      return toCopy
+
+    if null toCopy
+      then trace NoBlocksToCopyToImmDB
```

Comment:
Add a comment: this can't happen in practice because we're only called when the fragment is longer than `k`. However, in the q-s-m tests, we are calling this function manually and it *can* happen.